.teamcity/src/main/kotlin/model/FunctionalTestBucketGenerator.kt

```diff
@@ -0,0 +1,213 @@
+                )
+            }
+        }
+    }
+
+    // docs subproject is special
+    private fun splitDocsSubproject(allSubprojects: List<GradleSubproject>): List<BuildTypeBucket> {
+        val docs = allSubprojects.find { it.name == "docs" }!!
+        val docs1 = LargeSubprojectSplitBucket(docs, 1, true, listOf(TestClassTime("org.gradle.docs.samples.Bucket1SnippetsTest", "docsTest", -1)))
+        val docs2 = LargeSubprojectSplitBucket(docs, 2, true, listOf(TestClassTime("org.gradle.docs.samples.Bucket2SnippetsTest", "docsTest", -1)))
+        val docs3 = LargeSubprojectSplitBucket(docs, 3, true, listOf(TestClassTime("org.gradle.docs.samples.Bucket3SnippetsTest", "docsTest", -1)))
+        val docs4 = LargeSubprojectSplitBucket(
+            docs, 4, false, listOf(
+                TestClassTime("org.gradle.docs.samples.Bucket1SnippetsTest", "docsTest", -1),
+                TestClassTime("org.gradle.docs.samples.Bucket2SnippetsTest", "docsTest", -1),
+                TestClassTime("org.gradle.docs.samples.Bucket3SnippetsTest", "docsTest", -1)
+            )
+        )
+        return listOf(docs1, docs2, docs3, docs4)
+    }
+
+    private fun splitIntoBucketsExcludingSpecialBuckets(
+        specialSubprojectNames: List<String>,
+        validSubprojects: List<GradleSubproject>,
+        subProjectTestClassTimes: List<SubprojectTestClassTime>,
+        testCoverage: TestCoverage
+    ): List<BuildTypeBucket> {
+        val specialSubprojects = validSubprojects.filter { specialSubprojectNames.contains(it.name) }
+        val otherSubProjectTestClassTimes = subProjectTestClassTimes.filter { !specialSubprojectNames.contains(it.subProject.name) }
+        return splitIntoBuckets(
+            LinkedList(otherSubProjectTestClassTimes),
+            SubprojectTestClassTime::totalTime,
+            { largeElement: SubprojectTestClassTime, size: Int -> largeElement.split(size) },
+            { list: List<SubprojectTestClassTime> -> SmallSubprojectBucket(list) },
+            testCoverage.expectedBucketNumber - specialSubprojects.size,
+            MAX_PROJECT_NUMBER_IN_BUCKET
+        )
+    }
+
+    private fun determineSubProjectClassTimes(testCoverage: TestCoverage, buildProjectClassTimes: BuildProjectToSubprojectTestClassTimes): Map<String, List<TestClassTime>>? {
+        val testCoverageId = testCoverage.asId("Gradle_Master_Check")
+        return buildProjectClassTimes[testCoverageId] ?: if (testCoverage.testType == TestType.soak) {
+            null
+        } else {
+            val testCoverages = model.stages.flatMap { it.functionalTests }
+            val foundTestCoverage = testCoverages.firstOrNull {
+                it.testType == TestType.platform &&
+                    it.os == testCoverage.os &&
+                    it.buildJvmVersion == testCoverage.buildJvmVersion
+            }
+            foundTestCoverage?.let {
+                buildProjectClassTimes[it.asId("Gradle_Check")]
```

Comment:
```suggestion
                buildProjectClassTimes[it.asId("Gradle_Master_Check")]
```

:x: I suppose this is wrong. We need to extract the constant `MASTER_CHECK_CONFIGURATION ` again, right?

Reply:
Thanks, done.